File: udemy/ds-and-a/DSAndA/StacksAndQueues/Assignment2/TwoStacksUsingSingleArray.cs
namespace DSAndA.StacksAndQueues.Assignment2
{
    public class TwoStacksUsingSingleArray
    {
        private int maxSize1;
        private int maxSize2;
        private int[] twoStacksArray;
        private int top1;
        private int top2;

        public TwoStacksUsingSingleArray(int maxSize1, int maxSize2)
        {
            this.maxSize1 = maxSize1;
            this.maxSize2 = maxSize2;
            this.twoStacksArray = new int[maxSize1 + maxSize2];
            this.top1 = -1;
            this.top2 = maxSize2 - 1;
        }

        public void Push1(int element)
        {
            if (this.top1 == this.maxSize1 - 1) throw new System.ApplicationException("Stack 1 is full.");

            this.twoStacksArray[++this.top1] = element;
        }

        public int Pop1()
        {
            return this.twoStacksArray[this.top1--];
        }

        public void Push2(int element)
        {
            this.twoStacksArray[++this.top2] = element;
        }

        public int Pop2()
        {
            return this.twoStacksArray[this.top2--];
        }
    }
}

File: udemy/ds-and-a/javascript-solutions/src/StackAndQueues/Assignment2/TwoStacksUsingSingleArray.js
class TwoStacksUsingSingleArray {
    constructor(maxSize1, maxSize2) {
        this.stack = [];
        for (let i = 0; i < maxSize1 + maxSize2; i++) {
            this.stack[i] = 0;
        }

        this.maxSize1 = maxSize1;
        this.maxSize2 = maxSize2;

        this.head1 = -1;
        this.top1 = -1;

        this.head2 = -1 + this.maxSize1;
        this.top2 = -1 + this.maxSize1;
    }

    push1(data) {
        if (this.top1 === this.maxSize1 - 1) {
            throw new Error("Stack is full");
        }

        this.stack[++this.top1] = data;

        if (this.head1 < 0) {
            this.head++;
        }
    }

    peek1() {
        return this.stack[this.top1];
    }

    push2(data) {
        if (this.top2 === this.maxSize1 + this.maxSize2 - 1) {
            throw new Error("Stack is full");
        }
        this.stack[++this.top2] = data;

        if (this.head2 < 0 + this.maxSize1) {
            this.head++;
        }
    }

    peek2() {
        return this.stack[this.top2];
    }
}

module.exports = TwoStacksUsingSingleArray;